bitcoin/src/pow.rs

```diff
@@ -0,0 +1,1530 @@
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Work(U256);
+
+impl Work {
+    /// Lowest possible work value for Mainnet. See comment on [`Params::pow_limit`] for more info.
+    pub const MAINNET_MIN: Work = Work(U256(0x0000_0000_ffff_0000_0000_0000_0000_0000_u128, 0));
+
+    /// Lowest possible work value for Testnet. See comment on [`Params::pow_limit`] for more info.
+    pub const TESTNET_MIN: Work = Work(U256(0x0000_0000_ffff_0000_0000_0000_0000_0000_u128, 0));
+
+    /// Lowest possible work value for Signet. See comment on [`Params::pow_limit`] for more info.
+    pub const SIGNET_MIN: Work = Work(U256(0x0000_0377_ae00_0000_0000_0000_0000_0000_u128, 0));
+
+    /// Lowest possible work value for Regtest. See comment on [`Params::pow_limit`] for more info.
+    pub const REGTEST_MIN: Work = Work(U256(0x7fff_ff00_0000_0000_0000_0000_0000_0000_u128, 0));
+
+    /// Converts this [`Work`] to [`Target`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use bitcoin::Work;
+    /// let w = Work::from(0xdeadbeef_u32);
+    /// assert_eq!(w.to_target().to_work(), w)
+    /// ```
+    pub fn to_target(self) -> Target { Target(self.0.inverse()) }
+}
+do_impl!(Work);
+
+impl Add for Work {
+    type Output = Work;
+    fn add(self, rhs: Self) -> Self { Work(self.0.wrapping_add(rhs.0)) }
+}
+
+impl Sub for Work {
+    type Output = Work;
+    fn sub(self, rhs: Self) -> Self { Work(self.0.wrapping_sub(rhs.0)) }
+}
+
+/// A 256 bit integer representing target.
+///
+/// The SHA-256 hash of a block's header must be lower than or equal to the current target for the
+/// block to be accepted by the network. The lower the target, the more difficult it is to generate
+/// a block. (See also [`Work`].)
+///
+/// ref: <https://en.bitcoin.it/wiki/Target>
+///
+/// # Examples
+///
+/// ```
+/// use bitcoin::Target;
+/// let _ = Target::from(0xdeadbeef_u32);
+/// ```
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Target(U256);
+
+impl Target {
+    /// The maximum possible target.
+    ///
+    /// This value is used to calculate difficulty, which is defined as how difficult the current
+    /// target makes it to find a block relative to how difficult it would be at the highest
+    /// possible target. Remember highest target == lowest difficulty.
+    ///
+    /// ref: <https://en.bitcoin.it/wiki/Target>
+    // In Bitcoind this is ~(u256)0 >> 32 stored as a floating-point type so it gets truncated, hence
+    // the low 208 bits are all zero.
+    const MAX: Self = Target(U256(0xFFFF_u128 << (208 - 128), 0));
+
+    /// The maximum possible target (see [`Target::MAX`]).
+    ///
+    /// This is provided for compatibility with Rust 1.41.1, newer code should use [`Target::MAX`].
+    pub fn max_value() -> Self { Target::MAX }
```

Comment:
Rust 1.41.1 doesn't disallow using just `MAX`, so this isn't needed.

Reply:
Oh I thought it would be easier on devs writing code targeting 1.41.1 if they didn't have to use `max_value` sometimes and `MAX` sometimes. The comment "compatibility" is probably too strong. Can remove though if deemed unnecessary.

Reply:
Maybe say "consistency". (I only got the idea now, sorry.)

Reply:
Good idea, used as suggested.